### PR TITLE
ci: initial overhaul of tests

### LINKED ISSUE #88: Use Mock classes for testing
(where possible)

https://docs.pytest.org/en/stable/monkeypatch.html?highlight=mocking#monkeypatching-returned-objects-building-mock-classes

* It's probably fine to pull most of the 'mock' db layer out of the existing normalization tests, as long as there's a way to catch update requests before they alter a local DB, validate them, and prevent them from making any changes.

### LINKED ISSUE #177: Add mini data files to validate ETL methods
(and/or clean up current test data)

### LINKED ISSUE #306: Update test fixtures
Test fixtures are outdated and causes tests to fail. We should update the test fixtures with the latest data changes

### PR DESCRIPTION
**Pull request by jsstevenson**
close #177 
close #88  more or less
also close #306 

- Remove Tox (the experiment is over, I think it causes more headaches than it solves)
- Add a bunch of scripts to generate the various data needed to support tests in CI. They make use of the source ETL classes' data acquisition methods to pull from the newest version of things, which necessitated a few small tweaks.
- Everything seems to be performing correctly and I think I cleaned up all the WIP parts but apologies if I missed something
- In CI, some of the tests are ~dependent on each other (eg merge could fail if the individual source extractors do something wrong since it uses their side effects to produce normalized groups). I went back and forth on trying to hedge against this and decided that it's fine or at least not worth the trouble to do better. In the long term, I'd like to rewrite the `unit/` tests as true unit tests -- ie test inputs/outputs of individual ETL functions rather than checking that the entire pipeline output is correct -- and move all the existing tests to an `integration` or `end_to_end` subdirectory.
- I added some checks to refuse to run anything that might overwrite the DB in test if an AWS env var is active. I think I thought about this correctly.
 
Note for testing/pr reviewing:

- `export THERAPY_TEST=true` to use the minified data fixtures. This will overwrite data stored in your shared_local_db file, so proceed with caution.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -20,152 +44,73 @@ def test_data():
...
-            label_and_type = f"{query}##{match_type.lower()}"
-            records_lookup = self.records.get(label_and_type, None)
-            if records_lookup:
-                return [v.copy() for v in records_lookup.values()]
-            else:
-                return []
+    """Provide a database instance to be used by tests."""
+    database = Database()
+    if os.environ.get("THERAPY_TEST", "").lower() == "true":
+        if os.environ.get(AWS_ENV_VAR_NAME):
```
Comment (korikuzma) [inline tests/conftest.py:52]: Good catch. I should make an issue to do this for gene.

File: tests/conftest.py [hunk 2]
```diff
@@ -198,16 +143,23 @@ def compare_records(actual: Drug, fixt: Drug):
     if (actual.has_indication is not None) and (fixt.has_indication is not None):
         actual_inds = actual.has_indication.copy()
         fixture_inds = fixt.has_indication.copy()
-        assert len(actual_inds) == len(fixture_inds)
+        # assert len(actual_inds) == len(fixture_inds)
```
Comment (korikuzma) [inline tests/conftest.py:146]: Was this supposed to be added back? If not, go ahead and delete.

File: tests/data/rxnorm/rxnorm_2022-12-05.RRF [hunk 3]
```diff
@@ -0,0 +1,663 @@
+44|ENG||||||10283027||||USP|IN|m49500|Mesna||N||
```
Comment (korikuzma) [inline tests/data/rxnorm/rxnorm_2022-12-05.RRF:1]: Does the license allow us to share RxNorm data publicly?

File: tests/data/hemonc/hemonc_concepts_2022-08-21.csv [hunk 4]
```diff
@@ -0,0 +1,116 @@
+concept_name,vocabulary_id,concept_class_id,concept_code,valid_start_date,valid_end_date,invalid_reason
```
Comment (korikuzma) [inline tests/data/hemonc/hemonc_concepts_2022-08-21.csv:1]: Does the license allow us to share the Hemonc data publicly?

File: tests/scripts/build_ncit_data.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+
+ncit = NCIt(Database())
+ncit._extract_data()
+TEST_DATA_DIR = Path(__file__).resolve().parents[1] / "data" / "ncit"
+outfile_path = TEST_DATA_DIR / ncit._src_file.name
+
+
+def ncit_parser():
+    """TODO"""
```
Comment (korikuzma) [inline tests/scripts/build_ncit_data.py:35]: Remove todo

File: tests/data/rxnorm/rxnorm_2022-12-05.RRF (hunk 3 above)
Comment (jsstevenson) [reply tests/data/rxnorm/rxnorm_2022-12-05.RRF:1]: This is a good question and I'll probably have to double check w/ alex

File: tests/data/hemonc/hemonc_concepts_2022-08-21.csv (hunk 4 above)
Comment (jsstevenson) [reply tests/data/hemonc/hemonc_concepts_2022-08-21.csv:1]: It's CC BY, we should be fine

**Comment by jsstevenson** [comment]
todo: double check for consistency with https://github.com/cancervariants/gene-normalization/pull/159

File: tests/data/rxnorm/rxnorm_2022-12-05.RRF (hunk 3 above)
Comment (jsstevenson) [reply tests/data/rxnorm/rxnorm_2022-12-05.RRF:1]: I went ahead and pared down the file to just the rows in the permitted sources constant. I think that's a good enough middle ground.

File: tests/conftest.py [hunk 6]
```diff
@@ -20,152 +44,72 @@ def test_data():
...
-            if records_lookup:
-                return [v.copy() for v in records_lookup.values()]
-            else:
-                return []
+    """Provide a database instance to be used by tests."""
+    database = Database()
+    if os.environ.get("THERAPY_TEST", "").lower() == "true":
+        if os.environ.get(AWS_ENV_VAR_NAME):
+            assert False, (
+                f"Cannot have both GENE_TEST and {AWS_ENV_VAR_NAME} set."
```
Comment (korikuzma) [inline tests/conftest.py:54]:
```suggestion
                f"Cannot have both THERAPY_TEST and {AWS_ENV_VAR_NAME} set."
```

**Review by korikuzma — CHANGES_REQUESTED** [review]
Missed copy/paste change

File: tests/conftest.py (hunk 6 above)
Comment (jsstevenson) [reply tests/conftest.py:54]: lole

**Review by korikuzma — APPROVED** [review]

**Pull request merged**